README.md
Binkd is a Fidonet mailer designed to operate via TCP/IP networks.

As a FTN-compatible internet daemon, it makes possible efficient
utilization of TCP/IP protocol suite as a transport layer in
FTN-based (Fido Technology Network) networks.

## Compiling

non-UNIX:

1. Find in mkfls/ a subdirectory for your system/compiler, copy all files 
   to the root of the sources.
2. Run make (nmake, wmake or gmake, name of make's binary is rely with C
   compiler).

UNIXes:

1.) Clone the repo: 

`$ git clone https://github.com/pgul/binkd`

2.) Change into the new binkd source directory:

`$ cd binkd`

3.) Copy all files from mkfls/unix/ to the root of binkd sources:

`cp mkfls/unix/* .`

2.) Run configure and make:

`$ ./configure`
`$ make`

3.) When finished, the following instructions will be displayed offering various options for you:

```
 Binkd is successfully compiled.

 If you want to install Binkd files into /usr/local
     1. Run `make -n install' to be sure this makefile will
        do not something criminal during the installation;
     2. `su' to root;
     3. Run `make install' to install Binkd.
     4. Edit /usr/local/etc/binkd.conf-dist and RENAME it or
        MOVE it somewhere (so another `make install' will
        not overwrite it during your next Binkd upgrade)

 If you want to put the files into some other directory just
 run `configure --prefix=/another/path' and go to step 1.
```

## Installation

1. Edit sample binkd.cfg.
2. Run binkd.

## More info

**Echomail areas:**
* RU.BINKD (russian)
* BINKD (international)

**Web site:** http://www.corbina.net/~maloff/binkd/

**FTP:** ftp://happy.kiev.ua/pub/fidosoft/mailer/binkd/

**The mirrors:**
   * ftp://fido.thunderdome.us/pub/mirror/binkd/
   * ftp://cube.sut.ru/pub/mirror/binkd/
   * http://binkd.spb.ru

**Documentation:**

* [English manual for binkd 0.9.2](http://web.archive.org/web/20131010041927/http://www.doe.carleton.ca/~nsoveiko/fido/binkd/man/) © Nick Soveiko (<nsoveiko@doe.carleton.ca>)
* [Russian manual for binkd 0.9.9](http://binkd.grumbler.org/binkd-ug-ru.htm.win.ru) © Stas Degteff (`2:5080/102@fidonet`)
* [FAQ](http://binkd.grumbler.org/binkdfaq.shtml)

**Authors:** Dmitry Maloff <maloff@corbina.net> and others.

**Bug reporting:** <binkd-bugs@happy.kiev.ua>, also RU.BINKD or BINKD echoconferences.

**Binkd developers mailing list:** <binkd-dev@happy.kiev.ua> (send `subscribe binkd-dev` to <majordomo@happy.kiev.ua> for subscribe).


HISTORY
2022/07/02 13:06:13 1.1a-115 git
run.c,2.16,2.17
Set correct command line interpreter for Windows in fail situation
Signed-off-by: vasilyevmax <35378045+vasilyevmax@users.noreply.github.com>

2022/07/01 10:11:12 git
binkdfaq-en.txt,2.8,2.9 binkdfaq-ru.txt,2.8,2.9
update Max Vasilyev's site address

2022/06/25 08:55:47 1.1a-114 git
binkd.cfg,2.64,2.65 binkd.conf,1.9,1.10 ftnq.c,2.45,2.46 ftnq.h,2.12,2.13 readcfg.c,2.117,2.118 readcfg.h,2.45,2.46
Remove .try files after successful session

2022/05/19 20:31:12 git
binkd.cfg,2.63,2.64 binkd.conf,1.8,1.9
The -64 and -46 options and their descriptions have been added to the config.

2022/02/11 21:27:12 git
nodelist.pl,1.4,1.5
Fix some warnings in misc/nodelist.pl

2021/11/14 11:17:45 1.1a-113 git
protocol.c,2.238,2.239
Fix format in log and M_ERR messages

2021/07/08 10:01:11 git
Makefile.in,2.25,2.26 binkd.spec,1.214,1.215
fix binkd.spec and add a doc package with config and FAQ

2021/01/19 11:07:20 1.1a-112 git
breaksig.c,2.26,2.27
Graceful exit when idle for Windows builds, for Nick Andre.
Set the binkd_exit flag to TRUE when signaled to exit and return TRUE
(indicating that the signal was handled by the process) thus mimicking the
behavior of the *nix builds' handling of termination signals.

2021/01/19 11:07:11 git
Makefile,2.55,2.56
Define HAS_SNPRINTF for modern MSVC build compatibility (e.g. MSVC2019)
Fix build error using "nmake MSVC_VER=10" and MSVC2019:

Linking binkd.exe...
snprintf.obj : error LNK2005: _snprintf already defined in readcfg.obj
bin\msvc10-binkd\binkd.exe : fatal error LNK1169: one or more multiply defined symbols found

2020/05/26 20:48:08 1.1a-111 git
binkd.c,2.127,2.128
Print "Facilities" without space at the end of line. Make sense for further "binkd.exe -vv>file_id.diz" processing. (#21)

2020/05/25 14:12:23 git
configure,2.57,2.58 configure.in,2.57,2.58
Fix check perl multiplicity in configure

2020/05/25 08:48:41 1.1a-110 git
readcfg.c,2.116,2.117
Revert 4a528544 (perl multiplicity)

2020/05/17 23:32:09 1.1a-109 git
binkd.c,2.126,2.127 perlhooks.c,2.95,2.96 perlhooks.h,2.18,2.19 readcfg.c,2.115,2.116
Call PERL_SYS_INIT3() only once

2020/05/17 20:52:52 1.1a-108 git
exitproc.c,2.48,2.49
tab to spaces

2020/05/17 20:52:14 1.1a-107 git
readcfg.c,2.114,2.115
Require PERL_MULTIPLICITY for run multiple perl interpreters

2020/05/17 20:52:07 1.1a-106 git
perlhooks.c,2.94,2.95
Set PL_perl_destruct_level to 1 for correct restart perl interpreter
Signed-off-by: Pavel Gulchuk <gul@gul.kiev.ua>

2020/05/17 20:11:12 1.1a-105 git
perlhooks.c,2.93,2.94
Fix init perlsem

2020/04/30 21:30:12 git
Makefile,1.1,1.2
[DOS] fix making directory tree for DOS build

2020/04/29 09:03:10 1.1a-104 git
ftnnode.c,2.51,2.52 ftnq.c,2.44,2.45 protocol.c,2.237,2.238
Silence warnings: remove superfluous nullity checks
struct `FTN_NODE` defined in `btypes.h` contains
the member `pwd`, which is a character array.
Several places in the code, this array was checked
for nullity; however, this was always done in a place
where the validity of the referring `FTN_NODE` struct
pointer had already been established, and as `pwd` is
an array (and not a generic pointer) it can never be
NULL.  These superflous checks caused warnings when
compiling under `clang`.

This change removes those checks to silence these warnings.

Additionally, at those call sites often the code called
`strcmp` to compare `p->pwd` against some other string,
invariably inside of a conditional.  Modify the
conditionals to always explicity compare the value
returned from `strcmp` against e.g. 0.

Signed-off-by: Dan Cross <cross@fat-dragon.org>

2020/04/25 15:53:52 git
Makefile,2.28,2.29 Makefile.emo,1.11,1.12 Makefile.klibc,1.6,1.7
[OS/2] change .EXE naming scheme and library definitions

2020/04/25 15:53:47 1.1a-103 git
ftnq.c,2.43,2.44 run.c,2.15,2.16 sys.h,2.47,2.48
compile Binkd for DOS with djgpp 2.05 and watt32 library

2020/04/25 15:53:17 git
Makefile,NONE,1.1
compile Binkd for DOS with djgpp 2.05 and watt32 library

2020/04/25 15:53:10 1.1a-102 git
client.c,2.110,2.111
Fix build for OS/2 and Windows 9x on ancient compilers
Original patch by Dan Cross,
additional testing and improvements by Max Vasilyev
Neither Win9x nor OS/2 define `struct sockaddr_storage`.
To work around this, we resort to casts to make the
`ss_family` member of that struct look like the `sa_family`
member in a `struct sockaddr`.  According to POSIX, all
of these structs should be defined in such a way that
this will work without surprises.

2020/04/05 20:33:09 git
README.md,2.2,2.3
Update README.md
Fixed incorrect instructions and beautified the steps into a tested and verified cookbook style Howto.

2020/01/30 10:21:49 1.1a-101 git
client.c,2.109,2.110
Fix an out-of-bounds error on sockaddrs.
The `invalidAddresses` vector in `client.c` is used
to hold invalid addresses `binkd` should not use.
However, the array was of type `struct sockaddr`,
which is not large enough to hold all of the
protocol-specific data of the `struct sockaddr_*`
structures.  As a result, `binkd` would access
out-of-bounds memory when examining elements of
the array.

Fixed by redefining `invalidAddresses` to be an
array of type `struct sockaddr_storage`, which is
guaranteed by POSIX to be large enough to hold all
data associated with a socket address for any
protocol family.

Signed-off-by: Dan Cross <cross@fat-dragon.org>

2020/01/30 10:21:13 1.1a-100 git
client.c,2.108,2.109 ftnnode.c,2.50,2.51 iphdr.h,2.28,2.29 protocol.c,2.236,2.237 readcfg.c,2.113,2.114 readcfg.h,2.44,2.45
Fix use-after-free bug in get_host_and_port.
Don't use pointer assignment in this function,
but rather, copy into a fixed-length buffer.

Fixes #15

Signed-off-by: Dan Cross <cross@fat-dragon.org>

2019/01/23 09:16:18 git
binkdfaq-en.txt,2.7,2.8 binkdfaq-ru.txt,2.7,2.8
add a question on non-ASCII domains

2018/05/14 22:32:57 git
configure,2.56,2.57 configure.in,2.56,2.57
Updated autoconf scripts

2018/05/14 22:32:47 1.1a-99 git
readcfg.c,2.112,2.113

md5b.c
/* MD5C.C - RSA Data Security, Inc., MD5 message-digest algorithm
 */

/* Copyright (C) 1991-2, RSA Data Security, Inc. Created 1991. All
rights reserved.

License to copy and use this software is granted provided that it
is identified as the "RSA Data Security, Inc. MD5 Message-Digest
Algorithm" in all material mentioning or referencing this software
or this function.

License is also granted to make and use derivative works provided
that such works are identified as "derived from the RSA Data
Security, Inc. MD5 Message-Digest Algorithm" in all material
mentioning or referencing the derived work.

RSA Data Security, Inc. makes no representations concerning either
the merchantability of this software or the suitability of this
software for any particular purpose. It is provided "as is"
without express or implied warranty of any kind.

These notices must be retained in any copies of any part of this
documentation and/or software.
 */

#include <string.h>
#include <stdlib.h>
#include <ctype.h>
#include <stdio.h>
#ifdef HAVE_SYS_TIME_H
#include <sys/time.h>
#else
#include <time.h>
#endif
#include "sys.h"
#include "iphdr.h"
#include "protoco2.h"
#include "md5b.h"
#include "tools.h"
#include "server.h"

/* Constants for MD5Transform routine.
 */

#define S11 7
#define S12 12
#define S13 17
#define S14 22
#define S21 5
#define S22 9
#define S23 14
#define S24 20
#define S31 4
#define S32 11
#define S33 16
#define S34 23
#define S41 6
#define S42 10
#define S43 15
#define S44 21

static void MD5Transform PROTO_LIST ((UINT4 [4], unsigned char [64]));
static void Encode PROTO_LIST
  ((unsigned char *, UINT4 *, unsigned int));
static void Decode PROTO_LIST
  ((UINT4 *, unsigned char *, unsigned int));
static void MD5_memcpy PROTO_LIST ((POINTER, POINTER, unsigned int));
static void MD5_memset PROTO_LIST ((POINTER, int, unsigned int));

static unsigned char PADDING[64] = {
  0x80, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
};

/* F, G, H and I are basic MD5 functions.
 */
#define F(x, y, z) (((x) & (y)) | ((~x) & (z)))
#define G(x, y, z) (((x) & (z)) | ((y) & (~z)))
#define H(x, y, z) ((x) ^ (y) ^ (z))
#define I(x, y, z) ((y) ^ ((x) | (~z)))

/* ROTATE_LEFT rotates x left n bits.
 */
#define ROTATE_LEFT(x, n) (((x) << (n)) | ((x) >> (32-(n))))

/* FF, GG, HH, and II transformations for rounds 1, 2, 3, and 4.
Rotation is separate from addition to prevent recomputation.
 */
#define FF(a, b, c, d, x, s, ac) { (a) += F ((b), (c), (d)) + (x) + (UINT4)(ac); (a) = ROTATE_LEFT ((a), (s)); (a) += (b); }
#define GG(a, b, c, d, x, s, ac) { (a) += G ((b), (c), (d)) + (x) + (UINT4)(ac); (a) = ROTATE_LEFT ((a), (s)); (a) += (b); }
#define HH(a, b, c, d, x, s, ac) { (a) += H ((b), (c), (d)) + (x) + (UINT4)(ac); (a) = ROTATE_LEFT ((a), (s)); (a) += (b); }
#define II(a, b, c, d, x, s, ac) { (a) += I ((b), (c), (d)) + (x) + (UINT4)(ac); (a) = ROTATE_LEFT ((a), (s)); (a) += (b); }

/* MD5 initialization. Begins an MD5 operation, writing a new context.
 */
static void MD5Init (context)
MD5_CTX *context;                                        /* context */
{
  context->count[0] = context->count[1] = 0;
  /* Load magic initialization constants.
*/
  context->state[0] = 0x67452301;
  context->state[1] = 0xefcdab89;
  context->state[2] = 0x98badcfe;
  context->state[3] = 0x10325476;
}

/* MD5 block update operation. Continues an MD5 message-digest
  operation, processing another message block, and updating the
  context.
 */
static void MD5Update (context, input, inputLen)
MD5_CTX *context;                                        /* context */
unsigned char *input;                                /* input block */
unsigned int inputLen;                     /* length of input block */
{
  unsigned int i, index, partLen;

  /* Compute number of bytes mod 64 */
  index = (unsigned int)((context->count[0] >> 3) & 0x3F);

  /* Update number of bits */
  if ((context->count[0] += ((UINT4)inputLen << 3)) < ((UINT4)inputLen << 3))
   context->count[1]++;
  context->count[1] += ((UINT4)inputLen >> 29);

  partLen = 64 - index;

  /* Transform as many times as possible.
*/
  if (inputLen >= partLen) {
 MD5_memcpy
   ((POINTER)&context->buffer[index], (POINTER)input, partLen);
 MD5Transform (context->state, context->buffer);

 for (i = partLen; i + 63 < inputLen; i += 64)
   MD5Transform (context->state, &input[i]);

 index = 0;
  }
  else
 i = 0;

  /* Buffer remaining input */
  MD5_memcpy
 ((POINTER)&context->buffer[index], (POINTER)&input[i],
  inputLen-i);
}

/* MD5 finalization. Ends an MD5 message-digest operation, writing the
  the message digest and zeroizing the context.
 */
static void MD5Final (digest, context)
unsigned char digest[16];                         /* message digest */
MD5_CTX *context;                                       /* context */
{
  unsigned char bits[8];
  unsigned int index, padLen;

  /* Save number of bits */
  Encode (bits, context->count, 8);

  /* Pad out to 56 mod 64.
*/
  index = (unsigned int)((context->count[0] >> 3) & 0x3f);
  padLen = (index < 56) ? (56 - index) : (120 - index);
  MD5Update (context, PADDING, padLen);

  /* Append length (before padding) */
  MD5Update (context, bits, 8);

  /* Store state in digest */
  Encode (digest, context->state, 16);

  /* Zeroize sensitive information.
*/
  MD5_memset ((POINTER)context, 0, sizeof (*context));
}

/* MD5 basic transformation. Transforms state based on block.
 */
static void MD5Transform (state, block)
UINT4 state[4];
unsigned char block[64];
{
  UINT4 a = state[0], b = state[1], c = state[2], d = state[3], x[16];

  Decode (x, block, 64);

  /* Round 1 */
  FF (a, b, c, d, x[ 0], S11, 0xd76aa478); /* 1 */
  FF (d, a, b, c, x[ 1], S12, 0xe8c7b756); /* 2 */
  FF (c, d, a, b, x[ 2], S13, 0x242070db); /* 3 */
  FF (b, c, d, a, x[ 3], S14, 0xc1bdceee); /* 4 */
  FF (a, b, c, d, x[ 4], S11, 0xf57c0faf); /* 5 */
  FF (d, a, b, c, x[ 5], S12, 0x4787c62a); /* 6 */
  FF (c, d, a, b, x[ 6], S13, 0xa8304613); /* 7 */
  FF (b, c, d, a, x[ 7], S14, 0xfd469501); /* 8 */
  FF (a, b, c, d, x[ 8], S11, 0x698098d8); /* 9 */
  FF (d, a, b, c, x[ 9], S12, 0x8b44f7af); /* 10 */
  FF (c, d, a, b, x[10], S13, 0xffff5bb1); /* 11 */
  FF (b, c, d, a, x[11], S14, 0x895cd7be); /* 12 */
  FF (a, b, c, d, x[12], S11, 0x6b901122); /* 13 */
  FF (d, a, b, c, x[13], S12, 0xfd987193); /* 14 */
  FF (c, d, a, b, x[14], S13, 0xa679438e); /* 15 */
  FF (b, c, d, a, x[15], S14, 0x49b40821); /* 16 */

 /* Round 2 */
  GG (a, b, c, d, x[ 1], S21, 0xf61e2562); /* 17 */
  GG (d, a, b, c, x[ 6], S22, 0xc040b340); /* 18 */
  GG (c, d, a, b, x[11], S23, 0x265e5a51); /* 19 */
  GG (b, c, d, a, x[ 0], S24, 0xe9b6c7aa); /* 20 */
  GG (a, b, c, d, x[ 5], S21, 0xd62f105d); /* 21 */
  GG (d, a, b, c, x[10], S22,  0x2441453); /* 22 */
  GG (c, d, a, b, x[15], S23, 0xd8a1e681); /* 23 */
  GG (b, c, d, a, x[ 4], S24, 0xe7d3fbc8); /* 24 */
  GG (a, b, c, d, x[ 9], S21, 0x21e1cde6); /* 25 */
  GG (d, a, b, c, x[14], S22, 0xc33707d6); /* 26 */
  GG (c, d, a, b, x[ 3], S23, 0xf4d50d87); /* 27 */
  GG (b, c, d, a, x[ 8], S24, 0x455a14ed); /* 28 */
  GG (a, b, c, d, x[13], S21, 0xa9e3e905); /* 29 */
  GG (d, a, b, c, x[ 2], S22, 0xfcefa3f8); /* 30 */
  GG (c, d, a, b, x[ 7], S23, 0x676f02d9); /* 31 */
  GG (b, c, d, a, x[12], S24, 0x8d2a4c8a); /* 32 */

  /* Round 3 */
  HH (a, b, c, d, x[ 5], S31, 0xfffa3942); /* 33 */
  HH (d, a, b, c, x[ 8], S32, 0x8771f681); /* 34 */
  HH (c, d, a, b, x[11], S33, 0x6d9d6122); /* 35 */
  HH (b, c, d, a, x[14], S34, 0xfde5380c); /* 36 */
  HH (a, b, c, d, x[ 1], S31, 0xa4beea44); /* 37 */
  HH (d, a, b, c, x[ 4], S32, 0x4bdecfa9); /* 38 */
  HH (c, d, a, b, x[ 7], S33, 0xf6bb4b60); /* 39 */
  HH (b, c, d, a, x[10], S34, 0xbebfbc70); /* 40 */
  HH (a, b, c, d, x[13], S31, 0x289b7ec6); /* 41 */
  HH (d, a, b, c, x[ 0], S32, 0xeaa127fa); /* 42 */
  HH (c, d, a, b, x[ 3], S33, 0xd4ef3085); /* 43 */
  HH (b, c, d, a, x[ 6], S34,  0x4881d05); /* 44 */
  HH (a, b, c, d, x[ 9], S31, 0xd9d4d039); /* 45 */
  HH (d, a, b, c, x[12], S32, 0xe6db99e5); /* 46 */
  HH (c, d, a, b, x[15], S33, 0x1fa27cf8); /* 47 */
  HH (b, c, d, a, x[ 2], S34, 0xc4ac5665); /* 48 */

  /* Round 4 */
  II (a, b, c, d, x[ 0], S41, 0xf4292244); /* 49 */
  II (d, a, b, c, x[ 7], S42, 0x432aff97); /* 50 */
  II (c, d, a, b, x[14], S43, 0xab9423a7); /* 51 */
  II (b, c, d, a, x[ 5], S44, 0xfc93a039); /* 52 */
  II (a, b, c, d, x[12], S41, 0x655b59c3); /* 53 */
  II (d, a, b, c, x[ 3], S42, 0x8f0ccc92); /* 54 */
  II (c, d, a, b, x[10], S43, 0xffeff47d); /* 55 */
  II (b, c, d, a, x[ 1], S44, 0x85845dd1); /* 56 */
  II (a, b, c, d, x[ 8], S41, 0x6fa87e4f); /* 57 */
  II (d, a, b, c, x[15], S42, 0xfe2ce6e0); /* 58 */
  II (c, d, a, b, x[ 6], S43, 0xa3014314); /* 59 */
  II (b, c, d, a, x[13], S44, 0x4e0811a1); /* 60 */
  II (a, b, c, d, x[ 4], S41, 0xf7537e82); /* 61 */
  II (d, a, b, c, x[11], S42, 0xbd3af235); /* 62 */
  II (c, d, a, b, x[ 2], S43, 0x2ad7d2bb); /* 63 */
  II (b, c, d, a, x[ 9], S44, 0xeb86d391); /* 64 */

  state[0] += a;
  state[1] += b;
  state[2] += c;
  state[3] += d;

  /* Zeroize sensitive information.
*/
  MD5_memset ((POINTER)x, 0, sizeof (x));
}

/* Encodes input (UINT4) into output (unsigned char). Assumes len is
  a multiple of 4.
 */
static void Encode (output, input, len)
unsigned char *output;
UINT4 *input;
unsigned int len;
{
  unsigned int i, j;

  for (i = 0, j = 0; j < len; i++, j += 4) {
 output[j] = (unsigned char)(input[i] & 0xff);
 output[j+1] = (unsigned char)((input[i] >> 8) & 0xff);
 output[j+2] = (unsigned char)((input[i] >> 16) & 0xff);
 output[j+3] = (unsigned char)((input[i] >> 24) & 0xff);
  }
}

/* Decodes input (unsigned char) into output (UINT4). Assumes len is
  a multiple of 4.
 */
static void Decode (output, input, len)
UINT4 *output;
unsigned char *input;
unsigned int len;
{
  unsigned int i, j;

  for (i = 0, j = 0; j < len; i++, j += 4)
 output[i] = ((UINT4)input[j]) | (((UINT4)input[j+1]) << 8) |
   (((UINT4)input[j+2]) << 16) | (((UINT4)input[j+3]) << 24);
}

/* Note: Replace "for loop" with standard memcpy if possible.
 */

static void MD5_memcpy (output, input, len)
POINTER output;
POINTER input;
unsigned int len;
{
  unsigned int i;

  for (i = 0; i < len; i++)
 output[i] = input[i];
}

/* Note: Replace "for loop" with standard memset if possible.
 */
static void MD5_memset (output, value, len)
POINTER output;
int value;
unsigned int len;
{
  unsigned int i;

  for (i = 0; i < len; i++)
 ((char *)output)[i] = (char)value;
}

/* ---------------------------------------------------------- */
static void
hmac_md5(text, text_len, key, key_len, digest)
unsigned char*  text;                /* pointer to data stream */
int             text_len;            /* length of data stream */
unsigned char*  key;                 /* pointer to authentication key */
int             key_len;             /* length of authentication key */
MDcaddr_t         digest;              /* caller digest to be filled in */

{
        MD5_CTX context;
        unsigned char k_ipad[65];    /* inner padding -
                                      * key XORd with ipad
                                      */
        unsigned char k_opad[65];    /* outer padding -
                                      * key XORd with opad
                                      */
        unsigned char tk[16];
        int i;
        /* if key is longer than 64 bytes reset it to key=MD5(key) */
        if (key_len > 64) {

                MD5_CTX      tctx;

                MD5Init(&tctx);
                MD5Update(&tctx, key, key_len);
                MD5Final(tk, &tctx);

                key = tk;
                key_len = 16;
        }

        /*
         * the HMAC_MD5 transform looks like:
         *
         * MD5(K XOR opad, MD5(K XOR ipad, text))
         *
         * where K is an n byte key
         * ipad is the byte 0x36 repeated 64 times
         * opad is the byte 0x5c repeated 64 times
         * and text is the data being protected
         */

        /* start out by storing key in pads */
#ifndef bzero
#define bzero(x,y) memset((char*)x, 0, y)
#endif
#ifndef bcopy
#define bcopy(x,y,z) memmove((char*)y, (char*)x, z)
#endif
        bzero( k_ipad, sizeof k_ipad);
        bzero( k_opad, sizeof k_opad);
        bcopy( key, k_ipad, key_len);
        bcopy( key, k_opad, key_len);

        /* XOR key with ipad and opad values */
        for (i=0; i<64; i++) {
                k_ipad[i] ^= 0x36;
                k_opad[i] ^= 0x5c;
        }
        /*
         * perform inner MD5
         */
        MD5Init(&context);                   /* init context for 1st
                                              * pass */
        MD5Update(&context, k_ipad, 64);     /* start with inner pad */
        MD5Update(&context, text, text_len); /* then text of datagram */
        MD5Final(digest, &context);          /* finish up 1st pass */
        /*
         * perform outer MD5
         */
        MD5Init(&context);                   /* init context for 2nd
                                              * pass */
        MD5Update(&context, k_opad, 64);     /* start with outer pad */
        MD5Update(&context, digest, 16);     /* then results of 1st
                                              * hash */
        MD5Final(digest, &context);          /* finish up 2nd pass */
}
/* ---------------------------------------------------------- */
static void getrand(unsigned char *res, int len, STATE *rnd)
{
  MDcaddr_t digest;
  struct {
    time_t tm;
    unsigned short pid;
    unsigned short rand;
    int ext_rand;
  } rd;
  time(&rd.tm);
  rd.pid=(unsigned short)PID();
  rd.rand=rand();
  rd.ext_rand=ext_rand;
  hmac_md5((void *)&rd, sizeof(rd), (void *)rnd, sizeof(STATE), digest);
  if((rnd->peer_name)&&(rnd->peer_name[0]))
    hmac_md5((void *)rnd->peer_name, strlen(rnd->peer_name), digest, sizeof(digest), digest);
  memcpy(res, digest, len);
}

unsigned char *MD_getChallenge(char *str, STATE *rnd)
{
  unsigned char *res=NULL;
  int i;
  if(!str) {
    res=(unsigned char*)xalloc(MD_CHALLENGE_LEN+1);
    res[0]=MD_CHALLENGE_LEN;
    for(i=1;i<(MD_CHALLENGE_LEN+1);i+=MD5_DIGEST_LEN) 
    {
      int k=MD5_DIGEST_LEN;
      if(k>MD_CHALLENGE_LEN-i) k=MD_CHALLENGE_LEN-i;
      getrand(res+i, k, rnd);
    }
  }
  else {
    char *sp;
    if((sp=strstr(str, "CRAM"))==NULL) return NULL;
    if((sp=strstr(sp, "MD5"))==NULL) return NULL;
    while(sp[0]) if(*sp++=='-') break;
    if(!sp[0]) return NULL;
    for(i=0;isxdigit((int)sp[i]);i++) if(i>=128) break;
    i/=2;
    res=(unsigned char*)xalloc(i+1);
    res[0]=(char)i;
    for(i=0;isxdigit((int)sp[i]) && (i<128);i++) {
      unsigned char c=tolower(sp[i]);
      if(c>'9') c-='a'-10;
      else c-='0';
      if(!(i%2)) res[i/2+1]=c<<4;
      else res[i/2+1]|=c;
    }
  }	
  return res;
}

void MD_toString(char *rs, int len, unsigned char *digest)
{
  int i,j;
  if(!rs) return;
  strcpy(rs, "CRAM-MD5-");
  for(i=0,j=9;i<len;i++) 
  {
    unsigned char c=(digest[i]>>4);
    if(c>9) c+='a'-10;
    else c+='0';
    rs[j++]=c;
    c=(digest[i]&0xF);
    if(c>9) c+='a'-10;
    else c+='0';
    rs[j++]=c;
  }
  rs[j]=0;
}

char *MD_buildDigest(char *pw, unsigned char *challenge)
{
  char *rs=NULL;
  MDcaddr_t digest;
  if((!pw)||(!challenge)) return rs;
  hmac_md5(challenge+1, challenge[0], (void *)pw, strlen(pw), digest);
  rs=(char*)xalloc(MD5_DIGEST_LEN*2+10);
  MD_toString(rs, MD5_DIGEST_LEN, digest);
  return rs;
}
